File: daily/Binary Search/528-Random-pick-with-weight.py
import bisect
import random
from typing import List


class Solution:
    def __init__(self, w: List[int]):
        self.prefix = []
        running = 0
        for weight in w:
            running += weight
            self.prefix.append(running)
        self.total = self.prefix[-1]

    def pickIndex(self) -> int:
        target = random.randint(1, self.total)
        return bisect.bisect_left(self.prefix, target)

File: daily/Binary Search/528-Random-pick-with-weight.go
package daily

import "math/rand"

type Solution struct {
	prefix []int
	total  int
}

func Constructor(w []int) Solution {
	prefix := make([]int, len(w))
	prefix[0] = w[0]
	for i := 1; i < len(w); i++ {
		prefix[i] = prefix[i-1] + w[i]
	}
	return Solution{prefix, prefix[len(prefix)-1]}
}

func (s *Solution) PickIndex() int {
	target := rand.Intn(s.total) + 1 // random number from 1..total

	lo, hi := 0, len(s.prefix)-1
	for lo < hi {
		mid := (lo + hi) / 2
		if s.prefix[mid] < target {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}
